zdm/grid.py

```diff
@@ -0,0 +1,665 @@
+                Eth=self.thresholds[j,iz1,iDM2]*kz1 \
+                    + self.thresholds[j,iz2,iDM2]*kz2
+            else: 
+                Eth=self.thresholds[j,iz2,iDM2]*kDM2
+                Eth *= kz2**2 #assume threshold goes as Eth~z^2 in the near Universe
+        
+        # now account for beamshape
+        Eth /= MCb
+        
+        # NOW GET snr
+        #Eth=self.thresholds[j,k,l]/MCb
+        Es=np.logspace(np.log10(Eth),np.log10(self.Emax),100)
+        PEs=self.vector_cum_lf(Es,self.Emin,self.Emax,self.gamma)
+        PEs /= PEs[0] # normalises: this is now cumulative distribution from 1 to 0
+        r=np.random.rand(1)[0]
+        iE1=np.where(PEs>r)[0][-1] #returns list starting at 0 and going upwards
+        iE2 = iE1+1
+        # iE1 should never be the highest energy, since it will always have a probability of 0
+        kE1=(r-PEs[iE2])/(PEs[iE1]-PEs[iE2])
+        kE2=1.-kE1
+        MCE=10**(np.log10(Es[iE1])*kE1 + np.log10(Es[iE2])*kE2)
+        MCs=MCE/Eth
+        
+        FRBparams=[MCz,MCDM,MCb,j,MCs]
+        return FRBparams,pwb
+        
+
+    def update(self, vparams:dict, ALL=False):
+        """Update the grid based on a set of input
+        parameters
+        
+        Hierarchy:
+        Each indent corresponds to one 'level'.
+        This is used in the program control below
+        to dictate how far each tree should proceed
+        in calculation.
+        Direct variable inputs are always listed first
+        We see that sfr evolution and dm smearing
+        lie just before the pdv step
+        Hence, we deal with these first, and
+        calc rates as a final step regardless
+        of what else has changed.
+
+        Args:
+            vparams (dict):  dict containing the parameters
+                to be updated and their values
+            ALL (bool, optional):  If True, update the full grid
+        
+        calc_rates:
+            calc_pdv
+                Emin
+                Emax
+                gamma
+                calc_thresholds
+                    F0
+                    alpha
+                    bandwidth
+            set_evolution
+                sfr_n
+            
+            smear_grid
+                grid
+                mask
+                    dmx_params (lmean, lsigma)
+            dV
+            zdm_grid
+                H0
+        
+        Note that the grid is independent of the constant C (trivial dependence)
+
+        Args:
+            vparams (dict): [description]
+        """
+        # Init
+        reset_cos, get_zdm, calc_dV = False, False, False
+        smear_mask, smear_dm, calc_pdv, set_evol = False, False, False, False
+        new_sfr_smear, new_pdv_smear, calc_thresh = False, False, False
+
+        # Cosmology -- Only H0 so far
+        if self.chk_upd_param('H0', vparams, update=True):
+            reset_cos = True
+            get_zdm = True
+            calc_dV = True
+            smear_mask = True
```

Comment:
Changing H0: this does not require the line
smear_mask=True
since these are simply the host galaxy properties that depend only on z, and the mask should remain unchanged.